### ISSUE TITLE
autocompete with ajax

### ISSUE THREAD

**Issue by dellert** [Dogfalo/materialize#3539]
Is possible use autocompete plugin with ajax? For data load from server, something like this (but not woking):

```
$.ajax({
    url: $this.attr('data-url'),
    method: 'post',
    data: {
         city: $this.val()
    },
    dataType: 'json',
   success: function (res) {
          if (res != false) {
                $this.autocomplete({
                     data: res
                });
          }
     }
});
```

**Comment by dellert** [comment]
create pull request https://github.com/Dogfalo/materialize/pull/3541

**Comment by diegonc** [comment]
Here's another approach #3423 I've been thinking about. :smile: It's more general and moves custom logic outside of the component.

I'm not very fond of the solution using "classes" and I'd like to rework it in terms of simple functions some day. But I didn't have feedback yet, so didn't want to work advance any further without knowing if it's acceptable.

**Comment by dellert** [comment]
@Ketcap send you an email

**Comment by neokio** [comment]
This is so very necessary! The current autocomplete feature is incomplete. And the other options I've seen, including https://icefox0801.github.io/materialize-autocomplete/ , don't quite do it either.

Unlike a select (which can accommodate both a database id and visible value ... as in `<option value="51234">Blahblah</option>` ... the current autocomplete only shows the visible value, making it relatively useless for async database queries.

**Comment by timoschwarzer** [comment]
@neokio exactly. What we need is an implementation like this: https://www.devbridge.com/sourcery/components/jquery-autocomplete/

**Comment by neokio** [comment]
I came up with something that works pretty great. It requires `jquery.auto-complete.min.js` and `jquery.auto-complete.css` from https://github.com/Pixabay/jQuery-autoComplete ... which is beautifully coded, and only requires 3.4k.

Let's say you want an autocompleted city search input that shows the city name but returns the `city_id` ...

<b>HTML</b>:
```
	<div class="row">
		<div class="input-field">
			<input type="hidden" name="city_id" value="">
			<input id="city_autoselect" name="city_autoselect" type="text">
			<label for="city_autoselect">In which city do you live?</label>
		</div>
	</div>
```

<b>JS</b>:
```
	$('#city_autoselect').autoComplete({
		source: function(term, response){
			$.getJSON('/path/to/search.php', { q: term }, function(data){ response(data); });
		},
        renderItem: function (item, search){
            // escape special characters
            search = search.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
            var re = new RegExp("(" + search.split(' ').join('|') + ")", "gi");
            return '<div class="autocomplete-suggestion" data-id="'+ item[0] +'" data-val="' + item[1] + '">' + item[1].replace(re, "<b>$1</b>") + '</div>';
        },
		onSelect: function(e, term, item){
			$('#city_id').val( item.data('id') );
		}
	});
```
(The `renderItem` is almost identical to the default, just added `[0]` to the `item`. If, like me, you will only ever do this kind of autocomplete, you can just add `[0]` to the source code, and omit the `renderItem` from the JS call above.)


<b>PHP</b>:
```
	require '/path/to/requires.php';

	$res = [];
	$sql = $db->query("SELECT `id`,`name` FROM `cities` WHERE `name` LIKE '%". safeGET($_GET['q']) ."%' LIMIT 10");
	while ($r = $sql->fetch()) {
		$res[] = [$r['id'],$r['name']];
	}
	echo json_encode($res);
```

(Basically the PHP just outputs a JSON array like `[[255,'New York'], [1261,'New Orleans'], [8212,'Neward'], ...]`. Be sure to replace `safeGET()` with your own injection proofing methods.)

**Comment by yanickrochon** [comment]
+1 for asynchronous data sources

**Comment by nuplo** [comment]
check that code https://github.com/Dogfalo/materialize/pull/3694

**Comment by aashay0101** [comment]
@dellert Your modification to the existing Autocomplete Plugin is really amazing, but wouldn't it be better if instead of changing the existing plugin, you make another **autocompleteajax** function that can be used for ajax purposes and the normal **ajaxcomplete** function can be used for autocomplete purposes only. This way the existing code written using the **autocomplete** function wont have to be altered and would also ensure the backwards compatibility for versions 0.97.7 and below.

So basically adding your plugin just below the existing plugin (**autocomplete**) with different name like **autocompleteajax**

You also wont have to remove the existing Markup in forms.html file, but only add the documentation for the new plugin. You get the gist.

**Comment by yanickrochon** [comment]
@aashay0101 the term "ajax" is pretty much not used anymore. Also, many projects do not rely on jQuery's `$.ajax` function either. For example, with Meteor, the data source is from a `Collection`, using subscriptions. 

Personally, I would see this kind of API instead of the proposed one :

```js
$('#city_autocomplete').autoComplete({
  datasource: {
    query: function (term) {
      return someFunctionReturningPromise('cities', term);
    }
  },
  renderItem: function (item, search){
    // escape special characters
    search = search.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
    var re = new RegExp("(" + search.split(' ').join('|') + ")", "gi");
    return '<div class="autocomplete-suggestion" data-id="'+ item[0] +'" data-val="' + item[1] + '">' + item[1].replace(re, "<b>$1</b>") + '</div>';
  },
  onSelect: function(e, term, item){
    $('#city_id').val( item.data('id') );
  }
});
```

Returning a `Promise` implementation is becoming a standard, and even jQuery has implemented it since version 1.5, and now it is using the normalized [PromiseA](https://jquery.com/upgrade-guide/3.0/#breaking-change-and-feature-jquery-deferred-is-now-promises-a-compatible) implementation. Promises have the advantage of better handling asynchronous errors, and may be chained (no callback hell). And, since a promise is always returned, it does not matter if the resolved value is synchronous or asynchronous.

Bottom line is, do not use "ajax" anymore; we don't use much XML anymore for asynchronous communications :stuck_out_tongue:

**Comment by aashay0101** [comment]
@yanickrochon If you go through @dellert 's implementation it would break the websites which use the **.autocomplete** implementation, my suggestion was to separate both of them so that users can use any one of them depending on the need.

**Comment by yanickrochon** [comment]
@aashay0101 so, you base your suggestion to have 2 essentially same functionality components on a conflicting PR? Usually, PRs follow and address issues, not the other way around. Meaning that, before accepting a PR, one should discuss the issue.

In this case, a PR was made, and it is, de facto, not approved until it actually resolve all the concerning issues. And, in that sense, I say that

1. having two different components, essentially duplicating the functionality of each other, is not only redundant, but undesired in a framework
2. relying on the term "ajax" is not only limiting, but a bat design pattern; where features should not be based on a technology, but a functionality. The functionality, now, is to have **asynchronous data** not that the component should necessarily make the actually HTTP requests.

Therefore, the exact same component can be utilized for both situations seamlessly.

## Synchronous data

Providing data locally, using plain old JSON (i.e. an `Object`) (also, see #3586)

```js
$('input.autocomplete').autocomplete({
  data: {
    "Apple": null,
    "Microsoft": null,
    "Google": ...
  }
});
```

Where searching the input will simply gather results from the locally provided data. The actual underlying implementation can use a `Promise`, or not, it is irrelevant in the functionality.

## Asynchronous data

For example, using Meteor and [subscriptions](https://github.com/ccorcos/meteor-subs-cache)

```js
$('input.autocomplete').autocomplete({
  datasource: {
    query: function (term) {
      return new Promise(function (resolve) {
        Subs.subscribe('brands', term);
        Subs.onReady(function () {
          var data = {};
          Brands.find().forEach(function (brand) {
            data[ brand.name ] = null;
          }); 
          resolve( data );
        });
      });
    }
  }
});
```

Searching will update the autocomplete dropdown when the `Promise` is resolved with the subscribed documents (ex. ordered by relevance on the server). The idea of datasources can even be applied to, in fact, returned ordered lists of relevancy, and decouple that responsibility from Materialize. (Or Materialize could provide a basic datasource that would do the job...)

## Mixed data

Why not?

```js
$('input.autocomplete').autocomplete({
  data: {
    "Apple": null,
    "Microsoft": null,
    "Google": 'http://placehold.it/250x250'
  },
  datasource: {
    query: function (term) {
      return new Promise(function (resolve) {
        Subs.subscribe('brands', term);
        Subs.onReady(function () {
          var data = {};
          Brands.find().forEach(function (brand) {
            data[ brand.name ] = null;
          }); 
          resolve( data );
        });
      });
    }
  }
});
```

Both synchronous and asynchronous data can simply be merged and update the autocomplete dropdown. This can be done in a smart way.

## Conclusion

There is no need for two exact same components, especially when everything can preserve BC with no drawbacks. This feature can simply be a drop-in functionality in an application, which will require minimal refactor and trivial changes.

**Comment by dellert** [comment]
so, I rewrite and set better my autocomplete. But I will not create pull request, will use it just for my projects. If you don't want add normal  functionallity to your plugin, I'm very sorry.

**Comment by dellert** [comment]
Besides, materialize has many many troubles in functionallity and usage, but materialize developers ignore patches and pull requests from another programmers

**Comment by yanickrochon** [comment]
Why did you close this issue? The idea is good.

**Comment by dellert** [comment]
@yanickrochon because I found that materialize developers ignore patches and pull requests from another programmers:) I they want, they add self functional for this

**Comment by dellert** [comment]
@yanickrochon So, if you need, you can just replace autocomplete code with my:

````
/**************************
 * Auto complete plugin  *
 *************************/
$.fn.autocomplete = function (options) {
    // Defaults
    var defaults = {
        data: {},
        ajax: {},
        callback: null,
        delay: null,
        minLength: 2
    };

    options = $.extend(defaults, options);

    return this.each(function() {
        var $input = $(this);
        var data = options.data,
            $inputDiv = $input.closest('.input-field'); // Div to append on

        // Create autocomplete element
        var $autocomplete = $('<ul class="autocomplete-content dropdown-content"></ul>');

        // Append autocomplete element
        if ($inputDiv.length) {
            $inputDiv.append($autocomplete); // Set ul in body
        } else {
            $input.after($autocomplete);
        }

        var highlight = function(string, $el) {
            var img = $el.find('img');
            var matchStart = $el.text().toLowerCase().indexOf("" + string.toLowerCase() + ""),
                matchEnd = matchStart + string.length - 1,
                beforeMatch = $el.text().slice(0, matchStart),
                matchText = $el.text().slice(matchStart, matchEnd + 1),
                afterMatch = $el.text().slice(matchEnd + 1);
            $el.html("<span>" + beforeMatch + "<span class='highlight'>" + matchText + "</span>" + afterMatch + "</span>");

            if (img.length) {
                $el.prepend(img);
            }
        };

        var timer;

        // Perform search
        $input.on('keyup', function (e) {
            if($(this).val().length > options.minLength) {
                // Send ajax request
                if (timer){
                    clearTimeout(timer);
                }

                timer =  setTimeout(function() {
                    if(!$.isEmptyObject(options.ajax)) {
                        $.ajax({
                            url: options.ajax.url,
                            method: options.ajax.method,
                            data: {
                                data: $input.val(),
                                ajax_data: options.ajax.data
                            },
                            dataType: options.ajax.dataType,
                            error: options.ajax.error,
                            beforeSend: function(jqXHR, settings) {
                                options.ajax.beforeSend ? options.ajax.beforeSend(jqXHR, settings) : null;
                            },
                            success: function(res) {
                                data = res;

                                // If wanna work with response result
                                if(options.callback != null) {
                                    options.callback(res);
                                }
                            }
                        });
                    }

                    // Capture Enter
                    if (e.which === 13) {
                        $autocomplete.find('li').first().click();
                        return;
                    }
                }, options.delay);
            }

            var val = $input.val().toLowerCase();
            $autocomplete.empty();

            // Check if the input isn't empty
            if (val !== '') {
                $.each(data, function(i, value) {
                    if (value.value.toLowerCase().indexOf(val) !== -1 &&
                        value.value.toLowerCase() !== val) {
                        var autocompleteOption = $('<li data-id="'+ value.id +'"></li>');

                        if(!!value.image) {
                            autocompleteOption.append('<img src="'+ value.image +'" class="right circle"><span>'+ value.image +'</span>');
                        } else {
                            autocompleteOption.append('<span>'+ value.value +'</span>');
                        }

                        $autocomplete.append(autocompleteOption);
                        highlight(val, autocompleteOption);
                    }
                });
            }
        });

        // Set input value
        $autocomplete.on('click', 'li', function () {
            $input.val($(this).text().trim());

            if($('.autocomplete-id').length == 0) {
                $input.after("<input type='hidden' class='autocomplete-id' name='autocomplete-id' value='" + $(this).attr('data-id') + "'>");
            } else {
                $('.autocomplete-id').val($(this).attr('data-id'));
            }

            $autocomplete.empty();
        });
    });
};
````

Here you can send ajax & get data with ajax, set delay and min input value length if you need, can set custom data for send request & cal callback function if need

**Comment by yanickrochon** [comment]
When this line is executed : `data = res;` (in the request's success handler), the list of suggestion does not get automatically updated?

**Comment by dellert** [comment]
@yanickrochon oh! it's from old code:) updated

**Comment by ahmetalpbalkan** [comment]
Materialize currently lacks AJAX/XHR backed autocomplete functionality, which makes it a prototype-grade library. You can’t really load all the data in the browser in the real world applications. I am sad that this will be a deal-breaker for me...

**Comment by ngooide** [comment]
this project seems dying... can somebody else forge into another project and continue?

**Comment by taorg** [comment]
IMHO last two comments "lack" real criteria.
For sure there are tons of issues and potential improvements, but it doesn't mean it'd "dying".
And for sure, a back server communication "AJAX/XHR"  is not the scope of this project.

**Comment by dellert** [comment]
So, get it here https://github.com/dellert/autocompleteajax

**Comment by ahmetb** [comment]
Much appreciated! Thanks for creating a new repo. This will let folks send patches and make it better independently.

**Comment by dellert** [comment]
@ahmetb You can make it better:)

**Comment by bharathirasa** [comment]
Hi, i have a developed an ajax autocomplete using the devridge plugin. Get it using https://ampersandacademy.com/tutorials/materialize-css/materialize-css-framework-ajax-autocomplete-example-using-php

**Comment by fega** [comment]
Thanks @dellert to create this repo, I also believe that is a good idea make it independently.

**Comment by tomscholz** [comment]
I also added it [here](https://github.com/Dogfalo/materialize/wiki/Third-party-plugins)
